src/lib/SaferNames/PPrint.hs

```diff
@@ -0,0 +1,237 @@
+prettyAlt (Abs bs body) = hsep (map prettyBinderNoAnn  bs') <+> "->" <> nest 2 (p body)
+  where bs' = fromNest bs
+
+prettyBinderNoAnn :: Binder n l -> Doc ann
+prettyBinderNoAnn (b:>_) = p $ show b
+
+instance PrettyPrecE e => Pretty     (Abs Binder e n) where pretty = prettyFromPrettyPrec
+instance PrettyPrecE e => PrettyPrec (Abs Binder e n) where
+  prettyPrec (Abs binder body) = atPrec LowestPrec $ "\\" <> p binder <> "." <> pLowest body
+
+instance PrettyPrecE e => Pretty (PrimCon (e n)) where pretty = prettyFromPrettyPrec
+instance Pretty (PrimCon (Atom n)) where pretty = prettyFromPrettyPrec
+
+instance Pretty (Decl n l) where
+  pretty decl = case decl of
+    -- This is just to reduce clutter a bit. We can comment it out when needed.
+    -- Let (v:>Pi _)   bound -> p v <+> "=" <+> p bound
+    Let ann b rhs -> align $ p ann <+> p b <+> "=" <> (nest 2 $ group $ line <> pLowest rhs)
+
+instance Pretty (Atom n) where pretty = prettyFromPrettyPrec
+instance PrettyPrec (Atom n) where
+  prettyPrec atom = case atom of
+    Var v -> atPrec ArgPrec $ p v
+    Lam lamExpr@(LamExpr TabArrow _ _ _) ->
+      atPrec LowestPrec $ "\\for"
+      <+> prettyLamHelper lamExpr (PrettyLam TabArrow)
+    Lam lamExpr@(LamExpr arr _ _ _) ->
+      atPrec LowestPrec $ "\\"
+      <> prettyLamHelper lamExpr (PrettyLam arr)
+    Pi piType -> atPrec LowestPrec $ align $ prettyPiTypeHelper piType
+    TC  e -> prettyPrec e
+    Con e -> prettyPrec e
+    Eff e -> atPrec ArgPrec $ p e
+    DataCon name _ _ _ xs -> case xs of
+      [] -> atPrec ArgPrec $ p name
+      [l, r] | Just sym <- fromInfix (nameTag name) -> atPrec ArgPrec $ align $ group $
+        parens $ flatAlt " " "" <> pApp l <> line <> p sym <+> pApp r
+      _ ->  atPrec LowestPrec $ pAppArg (p name) xs
+    TypeCon name params -> case params of
+      [] -> atPrec ArgPrec $ p name
+      [l, r] | Just sym <- fromInfix (nameTag (getRawName name)) ->
+        atPrec ArgPrec $ align $ group $
+          parens $ flatAlt " " "" <> pApp l <> line <> p sym <+> pApp r
+      _  -> atPrec LowestPrec $ pAppArg (p name) params
+    LabeledRow items -> prettyExtLabeledItems items (line <> "?") ":"
+    Record items -> prettyLabeledItems items (line' <> ",") " ="
+    Variant _ label i value -> prettyVariant ls label value where
+      ls = LabeledItems $ case i of
+            0 -> M.empty
+            _ -> M.singleton label $ NE.fromList $ fmap (const ()) [1..i]
+    RecordTy items -> prettyExtLabeledItems items (line <> "&") ":"
+    VariantTy items -> prettyExtLabeledItems items (line <> "|") ":"
+    ACase e alts _ -> prettyPrecCase "acase" e alts
+    DataConRef _ params args -> atPrec AppPrec $
+      "DataConRef" <+> p params <+> p args
+    BoxedRef ptr size (Abs b body) -> atPrec AppPrec $
+      "Box" <+> p b <+> "<-" <+> p ptr <+> "[" <> p size <> "]" <+> hardline <> "in" <+> p body
+    ProjectElt idxs v ->
+      atPrec AppPrec $ "ProjectElt" <+> p idxs <+> p v
```

Comment:
Those things are very hard to understand, which is why we use `prettyProjection` for this. In general I am a bit afraid that we will end up loosing a few really good features during this change by missing small things like that.

Reply:
Though I guess at least for printing we might have some golden outputs in the tests...

Reply:
Definitely something to be aware of. As you say, the golden outputs should at least catch the user-facing stuff.